server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,233 @@
+import org.infinispan.rest.operations.exceptions.NoKeyException;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_JSON;
+import static org.infinispan.rest.framework.Method.DELETE;
+import static org.infinispan.rest.framework.Method.GET;
+import static org.infinispan.rest.framework.Method.POST;
+import static org.infinispan.rest.framework.Method.PUT;
+
+/**
+ * Protobuf schema manipulation Resource
+ *
+ * @author Katia Aresti
+ * @since 11
+ */
+public class ProtobufResource extends BaseCacheResource implements ResourceHandler {
+
+   private final ObjectMapper objectMapper;
+
+   public ProtobufResource(InvocationHelper invocationHelper) {
+      super(invocationHelper);
+      objectMapper = invocationHelper.getMapper();
+   }
+
+   @Override
+   public Invocations getInvocations() {
+      return new Invocations.Builder()
+            // Key related operations
+            .invocation().methods(GET).path("/v2/schema").handleWith(this::getSchemasNames)
+            .invocation().methods(POST).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,true))
+            .invocation().methods(PUT).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,false))
+            .invocation().methods(GET).path("/v2/schema/{schemaName}").handleWith(this::getSchema)
+            .invocation().method(DELETE).path("/v2/schema/{schemaName}").handleWith(this::deleteSchema)
+            .create();
+   }
+
+   private CompletionStage<RestResponse> getSchemasNames(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+
+      return CompletableFuture.supplyAsync(() ->
+          Flowable.fromIterable(cache.entrySet())
+               .filter(cacheEntry -> !((String)cacheEntry.getKey()).endsWith(ProtobufMetadataManager.ERRORS_KEY_SUFFIX))
+               .map(cacheEntry -> {
+                  String error = (String) cache
+                        .get(cacheEntry.getKey() + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX);
+                  ProtoSchema protoSchema = new ProtoSchema();
+                  protoSchema.name = cacheEntry.getKey().toString();
+                  if (error != null) {
+                     protoSchema.error = createErrorContent(protoSchema.name, error);
+                  }
+                  return protoSchema;
+               })
+                .collectInto(new HashSet<ProtoSchema>(), Set::add)
+                .map(protoSchemas -> {
```

Comment:
This seems a bit odd that we are collecting into a Set to then sort and then collect it into a List. can we not just use http://reactivex.io/RxJava/3.x/javadoc/io/reactivex/rxjava3/core/Flowable.html#sorted-java.util.Comparator- and then collect that into a List that is finally mapped to the response?

Reply:
yes